### PR TITLE
SSE2 to NEON compatibility

### PR DESCRIPTION
**Pull request by falkTX**
Everything builds, but I did not test them yet. :black_joker: 
This patch-set is used in the KXStudio repositories armhf and arm64 builds.

### CODE REVIEW

File: src/common/dsp/QuadFilterUnit.cpp [hunk 1]
```diff
@@ -691,7 +691,7 @@ __m128 TANH(__m128 in, __m128 drive)
    return _mm_max_ps(_mm_min_ps(y, y_max), y_min);
 }
 
-#if !_M_X64
+#if !_M_X64 && !defined(NEON_SSE)
```
Comment (baconpaul) [inline src/common/dsp/QuadFilterUnit.cpp]: These SSE1 instructions aren't actually used anywhere now anyway - I should really remove them.

File: src/linux/ConfigurationXml.S [hunk 2]
```diff
@@ -5,6 +5,8 @@ configurationXmlStart:
     .incbin "../../resources/data/configuration.xml"
 configurationXmlEnd:
     .global configurationXmlNullTerminator
+#ifndef __arm__
```
Comment (baconpaul) [inline src/linux/ConfigurationXml.S]: We will need this to be null terminated somehow otherwise the read will fail and surge can crash in some situations.

**Comment by baconpaul** [comment]
Tossed in a few comments. All looks pretty reasonable. Thank you for pushing these back upstream!

One other one is: sse2neon.h should be placed in libs/sse2neon with a clear license and provenance file (like we did with catch2) and not into src/common.

**Comment by falkTX** [comment]
No problem.
I did this as a test first to see if it was possible. So you think eventually it could be merged?
The xml assembly stuff I have no idea how to fix or handle.

**Comment by baconpaul** [comment]
Yeah I would have no problem merging this. In fact I would like to. Probably needs a bit of focus but not impossible at all. 

And we are sort of increasing our risk with the head of the codebase for a bit now we have 1.6.6 shipped and are pushing towards 1.7 - so this is a good time to do it.

**Comment by hexdump0815** [comment]
just for reference: i'm doing something similar (but more ugly i guess) on my armhf/aarch64 build of vcvrack - i use the more complete simde headers:
https://github.com/hexdump0815/vcvrack-dockerbuild-v1/blob/master/prepare.sh#L28
and then simply rename all the function call accordingly in a brute force way (the ugly part):
https://github.com/hexdump0815/vcvrack-dockerbuild-v1/blob/master/simde-ify.sh
is this maybe what you ment with xml stuff?
https://github.com/hexdump0815/vcvrack-dockerbuild-v1/blob/master/surge-rack-surge.armv7l.patch
and on armhf i had to disable unaligned access to avoid crashes:
https://github.com/hexdump0815/vcvrack-dockerbuild-v1/blob/master/surge-rack.armv7l.patch

best wishes - hexdump

**Comment by falkTX** [comment]
That is very helpful, thanks!
Going to see if the `%` operator works for my builds, I assume yes.

Any special reason to pick `-mfpu=vfpv3-d16` over the usual `-mfpu=neon-vfpv4` ?

**Comment by falkTX** [comment]
Sadly I still get a crash in the build server.
So I will have to run the armhf in actual hardware now for real testing :/

**Comment by baconpaul** [comment]
So all that is doing is null terminating the configuration xml memory block. Any other way to stick a zero at the end would work too

**Comment by falkTX** [comment]
I expected the `%object` to do the trick, since it seems to work for @hexdump0815 
But perhaps the issue is somewhere else, or the cloud builder environment has missing instructions.
I will have more details soon, need to find and install debian on my Pi. I am not in a rush though

**Comment by baconpaul** [comment]
One other thought: linux and only linux reads that configuration from memory. on mac and windows we read it from resources. If you have a #if ARM you could modify SurgeStorage.cpp (you can clearly see where it gets it) and add an if dev to read from resources.

I think, in retrospect, it was a mistake to do linux this way but back when we all were doing it it seemed reasonable. Since we need so many other things from the shared resource directory though singling this one out on only linux is something we could reverse if it becomes a real problem

**Comment by hexdump0815** [comment]
@falkTX - in case you would like to have a ready to go debian or ubuntu image for a raspberry-pi 3 - you might use those: https://github.com/hexdump0815/imagebuilder/releases/tag/200206-01 ... they are available in armhf and aarch64 versions :)

and here are some useful audio optimizaions for the rpi (they are for raspbian, but should be working on debian and ubuntu too): https://github.com/hexdump0815/vcvrack-dockerbuild-v1/blob/master/readme-raspbian.txt ... in that context maybe interesting too: https://github.com/hexdump0815/sonaremin/blob/master/files/extra-files/etc/rc.audio

good luck and best wishes - hexdump

**Comment by baconpaul** [comment]
Is there a docker image or some such we can use to try a pi build?

**Comment by hexdump0815** [comment]
@baconpaul - maybe this might help you to get going: https://github.com/hexdump0815/vcvrack-dockerbuild-v1/blob/master/readme-arm-build-on-intel.txt

**Comment by hexdump0815** [comment]
@baconpaul @falkTX - if any of you should have any compiled builds for armhf and aarch64 ready and want me to test them, please let me know - i can easily do so on ubuntu 18.04 and debian buster on both armhf and aarch64 - just let me know ...

**Comment by baconpaul** [comment]
Hi @hexdump0815 sorry I haven't actually been active on this - been working on skinning engine for a 1.7 release. Nothing coming from me.

**Comment by hexdump0815** [comment]
@baconpaul - no problem, there is no hurry :)

**Comment by hexdump0815** [comment]
ok - i gave the hacking approach i used for the surge-rack module for vcvrack a try to compile the surge plugin itself that way too and finally got it working on both armv7l and aarch64 ... this approach is really just a hack to get it compiled, but its at least a proof of concept that surge works on those platforms - i tested the vst's with reaper on both of them ... here are my notes and patches from this try: https://github.com/hexdump0815/surge-arm-build ... under releases one can find compiled versions ...

**Comment by baconpaul** [comment]
Wow simdeify.sh has quite a regex!!

This is cool. It looks like mostly flag changes. I wonder if we could coax premake, if an ARM variable is set, to generate those diffs without a patch. So basically have premake make a makefile which runs simdyfy and outputs the changed flags.

Thank you for sharing that. Neat stuff.

**Comment by hexdump0815** [comment]
@baconpaul - i played around a bit more with surge in reaper on armv7l and it looks like this architecture (i tried it on a rockchip rk3288, which already is one of the faster armv7l  cpus) is maybe a bit weak for surge - as soon as i use more complex patches (like "back to tyrell" from noisegeek) it quickly results in dropouts as soon as i play polyphone notes and somehow reaper/libjack and qjackctl both start to saturate a cpu each ... i only see this with surge there - helm or zynaddsubfx for instance do not show this phenomenon ... are there any tricks or changes (maybe even compile time) to make surge a bit less cpu hungry on such low end platforms (maybe increasing some internal buffers or such)?

a lot of thanks in advance and best wishes - hexdump

**Comment by baconpaul** [comment]
So the trick is really “use the SSE2 instruction set!” When you substitute it out what are you replacing it with? The synth is super duper CPU efficient on x86/SSE2 architectures precisely because of all the hand coded SSE2 and if you replaced that with something less optimal it may destroy performance.

That said I have no idea how powerful armv71 is. Surge was originally written to run OK on 2005 era intel chips. How does the armv71 compare?

**Comment by hexdump0815** [comment]
it should be in that cpu power range, maybe even a bit faster - its similar to the 2x1.6ghz sl7500 of the thinkpad x61s i'm writing this on right now - its a 4x1.8ghz 32bit cpu and it has something simd comparable to sse2 called neon and this simde stuff is trying to translate sse2 as good as automatically possible to neon which will not be as good as hand coded for sure (maybe with the compiler options i'm using to avoid the unaligned access crashes the neon instructions are not yet fully utilized even) ... at least on the 64bit aarch64 arm it seems to work pretty well - there it is pretty useable on an odroid c2 for instance (4x1.5ghz arm a53 64bit cores) ... all this is no reason to worry, its at least running to some degree even on armv7l and maybe someone with good arm know how will come around and has some good ideas to improve it further even on this dying species of older 32bit arm cpus :)

**Comment by baconpaul** [comment]
Yeah so the last time I thought about a non-x86 instruction set was very very long ago so I am way out of my league. But your comment on alignment is one that perhaps matters. Surge makes an effort to align blocks on 16 bit boundaries (of course - SSE2 requires that) using the C++-11 and C++-14 alignas directive. Does your compiler toolchain obey that? On win 32 bit with vs2017 for instance to make it work I had to turn on the C++-17 flag (even though alignas is a C++-11 and 14 directive). (Win 64 bit, linux, and mac were fine with alignment directives without the 17 uptick).

I don’t know if any of this helps, but I would like to try and help.

Amazing that it is working on the other chipset. Do you think somehow that it is mergable? I would be thrilled to try and do so.

**Comment by hexdump0815** [comment]
that c++17 seems to partially apply to the armv7l gcc as well: i recompiled with c++14 replaced with c++17 and with this the lv2 linking does not crash anymore and i can scan the vst3 plugin properly in reaper, but as soon as i send some midi to it it still crashes with a bus error due to unaligned access.

so we have the choice between "-march=armv7-a -mfpu=vfpv3-d16 -mno-unaligned-access" which does not seem to use neon simd and thus is slow  but does not crash or "-march=armv7-a -mfpu=neon -mno-unaligned-access" which would most probably be faster as it would use neon simd but crashes due to unaligned access - so we have an alignment problem somewhere on this platform: either in the surge code or in the simde code.

i tried to build the vst3 with debug info and run reaper from gdb, but this did not give any more details where it crashed. interesting was that i first replaced "-O3" with "-g" and then reaper crashed already on startup (most probably during plugin scan), putting back "-O3" as well made it start and crash only on midi input - so looks like all this is quite fragile. looks like here someone with deep arm and neon know how would be required.

**Comment by baconpaul** [comment]
So I think the question is whether alginas works properly. Lemme see if I can dig up the teensy test program (or recreate it) I used when doing 32 bit windows

**Comment by baconpaul** [comment]
Oh smart I put it all in a GitHub issue

https://github.com/surge-synthesizer/surge/issues/413

Can you take the very very last version of my test program in that issue and run it in your various flag environments with your compile tool chain? That issue explains the problem we faced basically.  Thanks!!

**Comment by hexdump0815** [comment]
i can only run your first test program, as the second uses _aligned_malloc and _aligned_free which seems to be a microsoft only thing ... independent of compiler options i always get the following on armv7l:
```
SizeOf size_t is 4
SizeOf void * is 4

Running test for unAl alignof=1
16 bit alignments is 344/1000

Running test for unAlSub alignof=1
16 bit alignments is 280/1000

Running test for al alignof=16
16 bit alignments is 1000/1000
```
on aarch64 i get:
```
SizeOf size_t is 8
SizeOf void * is 8

Running test for unAl alignof=1
16 bit alignments is 1000/1000

Running test for unAlSub alignof=1
16 bit alignments is 1000/1000

Running test for al alignof=16
16 bit alignments is 1000/1000
```
does that tell you anything in any way?

**Comment by baconpaul** [comment]
Right so that means that arm71 is ignoring the c++ alignas directive 
That’s why you get unaligned memory
And that’s why you crash without noalign
And that’s why you are slow without aligned neon

So cool now we know the problem! Now off to google...

(By the way one solution is to in an ifdef arm71 overload opnew perhaps so not all hope is lost but it may be real chunk of work)

**Comment by baconpaul** [comment]
oh i'm sorry ignore me. i mis-read the program output and alignment *is* working on arm71
harumph
sorry lots going on and was trying to answer between other things.

**Comment by schnitzeltony** [comment]
Have a question related the original PR:

Why do you suggest NEON_SSE (guess it is intended to be set by build-system) and don't use __ARM_NEON__ which is set automatically by compiler?

**Comment by schnitzeltony** [comment]
Aargh - cmark magic - I meant

\_\_ARM_NEON__

**Comment by baconpaul** [comment]
I just last week moved master from premake to cmake; I wonder if that would make some of these ports easier (for instance, we could avoid all the substitution steps and make a header which was included only in a neon build; merge that into master; etc...)

But again I don’t have the hardware or the know how to work on this. Happy to help though!

**Comment by schnitzeltony** [comment]
> I just last week moved master from premake to cmake; I wonder if that would make some of these ports easier (for instance, we could avoid all the substitution steps and make a header which was included only in a neon build; merge that into master; etc...)

I saw that after writing here. FWIW this PR does not apply any more.
Am working with cross environment (Yocto) and dozens of Raspis to test. Hope I'll find some time to rebase+rework this to master. It is worth because main coding effort was done by @falkTX.

**Comment by baconpaul** [comment]
Great! That’s cool.

And yeah also @hexdump0815 got it building in the rack-version (which contains all the dsp but none of the UI) and it worked fine in his arm for vcv rack. So definitely workable. Just would love to figure out what to merge back into the master branch and (aspirationally) if there’s a way I could have our pipelines test it still works somehow.

I have a mac and parallels. If yocto makes it easy for me to run your env under a docker container or in a linux vm happy to lend a hand.

But super happy to do a merge. We have continuous integration running to make sure mac win and linux keep working on any changes you suggest.

Thanks!

**Comment by hexdump0815** [comment]
@baconpaul - yes i think that header solution would be the right one and it could then easily be used for both native surge and surge-rack ... its just a matter of time to build that header file, which will be quite a bit of hand work i guess ... i was thinking of starting with my replace script and then use the diff as a base for which functions need to be mapped in such a header file - this way it should at least be possible to half automate the process ...

**Comment by baconpaul** [comment]
OK I want to at least do something with this PR before 1.7.0 so I am right now merging it into our cmake and testing it with the ubuntu cross compiler chain, using ARM_NEON and so on. Will update the readme and I hope merge. Probably will end up closing this PR and adding a new one - will update here when I do.

**Comment by baconpaul** [comment]
Closing in favor of #2342.

**Pull request closed without merging**